### PR TITLE
Validate location (offset) of the accumulated metadata when comparing

### PR DESCRIPTION
**Pull request by e4t**
Initially, the accumulated metadata location is initialized to HADDR_UNDEF
- the highest available address. Bogus input files may provide a location or size matching this value. Comparing this address against such bogus values may provide false positives. This make sure, the value has been initilized or fail the comparison early and let other parts of the code deal with the bogus address/size.
Note: To avoid unnecessary checks, we have assumed that if the 'dirty' member in the same structure is true the location is valid.

This fixes CVE-2018-13867 / Bug #2230.

Signed-off-by: Egbert Eich <eich@suse.com>

### CODE REVIEW

File: src/H5Faccum.c [hunk 1]
```diff
@@ -47,6 +47,7 @@
 #define H5F_ACCUM_THROTTLE  8
 #define H5F_ACCUM_THRESHOLD 2048
 #define H5F_ACCUM_MAX_SIZE  (1024 * 1024) /* Max. accum. buf size (max. I/Os will be 1/2 this size) */
+#define H5F_LOC_VALID(x)    (x != HADDR_UNDEF)
```
Comment (bmribler) [inline src/H5Faccum.c]: Is there a reason H5F_addr_defined(X) couldn't be used?

Comment (e4t) [reply src/H5Faccum.c]: No at all - except that I wasn't aware of this macro.  I'll change it.

Comment (bmribler) [reply src/H5Faccum.c]: Thanks!

**Review by jhendersonHDF — APPROVED** [review]

**Review by bmribler — APPROVED** [review]

File: src/H5Faccum.c [hunk 2]
```diff
@@ -484,7 +485,7 @@ H5F__accum_write(H5F_shared_t *f_sh, H5FD_mem_t map_type, haddr_t addr, size_t s
                     accum->size += size;
                 } /* end if */
                 /* Check if the piece of metadata being written overlaps the metadata accumulator */
-                else if (H5F_addr_overlap(addr, size, accum->loc, accum->size)) {
+                else if (H5F_addr_defined(accum->loc) && H5F_addr_overlap(addr, size, accum->loc, accum->size)) {
```
Comment (lrknox) [inline src/H5Faccum.c]:
```suggestion
                else if (H5F_addr_defined(accum->loc) &&
                         H5F_addr_overlap(addr, size, accum->loc, accum->size)) {
```

**Review by lrknox — APPROVED** [review]

File: release_docs/RELEASE.txt [hunk 3]
```diff
@@ -149,8 +149,21 @@ Bug Fixes since HDF5-1.13.3 release
 ===================================
     Library
     -------
-    -
+    - Fix CVE-2018-13867 / GHSA-j8jr-chrh-qfrf / Bug #2230
```
Comment (lrknox) [inline release_docs/RELEASE.txt]:
```suggestion
    - Fix CVE-2018-13867 / GHSA-j8jr-chrh-qfrf
```

File: release_docs/RELEASE.txt [hunk 4]
```diff
@@ -149,8 +149,21 @@ Bug Fixes since HDF5-1.13.3 release
...
+      Initially, the accumulated metadata location is initialized to HADDR_UNDEF
+      - the highest available address. Bogus input files may provide a location
+      or size matching this value. Comparing this address against such bogus
+      values may provide false positives. Thus make sure, the value has been
+      initilized or fail the comparison early and let other parts of the
+      code deal with the bogus address/size.
+      Note: To avoid unnecessary checks, it is assumed that if the 'dirty'
+      member in the same structure is true the location is valid.
+
+      (EFE - 2022/10/10 GH-2231)
```
Comment (lrknox) [inline release_docs/RELEASE.txt]:
```suggestion
      (EFE - 2022/10/10 GH-2230)
```

File: release_docs/RELEASE.txt [hunk 5]
```diff
@@ -149,8 +149,21 @@ Bug Fixes since HDF5-1.13.3 release
...
-    -
+    - Fix CVE-2018-13867 / GHSA-j8jr-chrh-qfrf / Bug #2230
  
+      Validate location (offset) of the accumulated metadata when comparing.
+
+      Initially, the accumulated metadata location is initialized to HADDR_UNDEF
+      - the highest available address. Bogus input files may provide a location
+      or size matching this value. Comparing this address against such bogus
+      values may provide false positives. Thus make sure, the value has been
+      initilized or fail the comparison early and let other parts of the
```
Comment (lrknox) [inline release_docs/RELEASE.txt]:
```suggestion
      initialized or fail the comparison early and let other parts of the
```

**Review by byrnHDF — APPROVED** [review]

**Pull request merged**